File: Commands/BaseScriptableCommand.cpp
//
// Copyright (c) 2010-2017 Fabric Software Inc. All rights reserved.
//

#include "CommandArgHelpers.h"
#include "BaseScriptableCommand.h"
#include <FabricUI/Application/FabricException.h>

using namespace FabricUI;
using namespace Commands;
using namespace Application;

BaseScriptableCommand::BaseScriptableCommand() 
  : BaseCommand()
{
}

BaseScriptableCommand::~BaseScriptableCommand() 
{
}

void BaseScriptableCommand::declareArg(
  const QString &key, 
  int flags, 
  const QString &defaultValue) 
{
  if(key.isEmpty())
    FabricException::Throw(
      "BaseScriptableCommand::declareArg",
      "declaring arg of '" + getName() + "', key not specified");

  ScriptableCommandArgSpec spec;
  spec.flags = flags;
  spec.defaultValue = defaultValue;
  m_argSpecs[key] = spec;

  if(!defaultValue.isEmpty())
    setArg(key, defaultValue);
}

bool BaseScriptableCommand::hasArg(
  const QString &key)
{
  return m_argSpecs.count(key) > 0;
}

bool BaseScriptableCommand::isArg(
  const QString &key,
  int flag)
{
  if(key.isEmpty()) 
    FabricException::Throw(
      "BaseScriptableCommand::isArg",
      "setting arg of '" + getName() + "', key not specified");

  if(!hasArg(key)) 
    // TODO: make this an optional behavior
    FabricException::Throw(
      "BaseScriptableCommand::isArg",
      "setting arg: '" + key + + "' not supported by command '" + getName() + "'");

  return (m_argSpecs[key].flags & flag);
}

QString BaseScriptableCommand::getArg(
  const QString &key)
{
  return m_args.count(key) > 0 
    ? m_args[key]
    : QString();
}

QList<QString> BaseScriptableCommand::getArgKeys()
{
  return m_argSpecs.keys(); 
}

bool BaseScriptableCommand::isArgSet(
  const QString &key)
{
  return m_args.count(key) && 
        !m_args[key].isEmpty();
}

void BaseScriptableCommand::setArg(
  const QString &key, 
  const QString &value) 
{
  if(key.isEmpty()) 
    FabricException::Throw(
      "BaseScriptableCommand::setArg",
      "setting arg of '" + getName() + "', key not specified");

  if(!hasArg(key)) 
    // TODO: make this an optional behavior
    FabricException::Throw(
      "BaseScriptableCommand::setArg",
      "setting arg: '" + key + + "' not supported by command '" + getName() + "'");

  m_args.insert(key, value);
}

void BaseScriptableCommand::validateSetArgs()
{
  QMapIterator<QString, ScriptableCommandArgSpec> it(m_argSpecs);
  
  while(it.hasNext()) 
  {
    it.next();

    QString key = it.key();
    ScriptableCommandArgSpec spec = it.value();
     
    if(!isArg(key, CommandArgFlags::OPTIONAL_ARG) && !isArgSet(key)) //is null
      FabricException::Throw(
        "BaseScriptableCommand::validateSetArgs",
        "validating arg: '" + key + "' of command '" + getName() + "' has not been set");
  }
}

QString BaseScriptableCommand::getArgsDescription() 
{
  QString res;

  FABRIC_CATCH_BEGIN();

  int count = 0;
  QMapIterator<QString, ScriptableCommandArgSpec> it(m_argSpecs);
  while(it.hasNext()) 
  {
    it.next();
    QString key = it.key();
    ScriptableCommandArgSpec spec = it.value();

    res += "    ["  + key + "]";
    res += ", opt: " + CommandArgHelpers::getArgsTypeSpecs(this, key);
    res +=  ", val: " + getArg(key);

    if(!spec.defaultValue.isEmpty())
      res += ", defVal: " + spec.defaultValue;
 
    res += (count < m_argSpecs.size() - 1) ? "\n" : "";
    count++;
  }

  FABRIC_CATCH_END("BaseScriptableCommand::getArgsDescription");

  return res;
}

File: Python/Canvas/Commands/BaseScriptableCommand.py
#
# Copyright (c) 2010-2017 Fabric Software Inc. All rights reserved.
#

from FabricEngine.FabricUI import Commands as CppCommands

class BaseScriptableCommand(CppCommands.BaseScriptableCommand_Python):
    
    """ BaseScriptableCommand specializes BaseScriptableCommand_Python (C++). 
        See Commands/BaseScriptableCommand.h(cpp), and Commands/BaseScriptableCommand_Python.h(cpp). 
    """
    
    def __init__(self):
        """ Initializes the BaseScriptableCommand.
        """
        super(BaseScriptableCommand, self).__init__()

    def declareArg(self, key, flag=2, defaultValue=""):
        """ Impl. of CppCommands.ScriptableCommand
            Declares an argument, called from constructor.
        """
        self._declareArg_Python(key, flag, defaultValue)
    
    def hasArg(self, key):
        """ Impl. of CppCommands.ScriptableCommand
            Checks if the command has an arg.
        """ 
        return super(BaseScriptableCommand, self).hasArg(key)
 
    def isArg(self, key):
        """ Impl. of CppCommands.ScriptableCommand
            Checks if an arg is optional.
        """
        error, hasFlag = self._isArg_Python(key)
        return hasFlag;

    def isArgLoggable(self, key):
        """ Impl. of CppCommands.ScriptableCommand
            Checks if an arg is loggable.
        """
        error, isLoggable = self._isArgLoggable_Python(key)
        return isLoggable;

    def getArgKeys(self):
        """ Impl. of CppCommands.ScriptableCommand
            Gets the arguments keys.
        """
        return super(BaseScriptableCommand, self).getArgKeys()
    
    def isArgSet(self, key):
        """ Impl. of CppCommands.ScriptableCommand
            Checks if an arg has been set, to get safely optional arg.
        """
        return super(BaseScriptableCommand, self).isArgSet(key)

    def getArg(self, key):
        """ Impl. of CppCommands.ScriptableCommand
            Gets an argument.
        """
        error, arg = self._getArg_Python(key) 
        return arg;

    def setArg(self, key, arg):
        """ Impl. of CppCommands.ScriptableCommand
            Sets an argument, called from the manager.
        """
        self._setArg_Python(key, arg)

    def validateSetArgs(self):
        """ Impl. of CppCommands.ScriptableCommand
            Checks the args are correctly set before doing the command, 
            called from the manager.
        """
        self._validateSetArgs_Python()

    def getArgsDescription(self):
        """ Impl. of CppCommands.ScriptableCommand
            Gets a decription of the args, for debugging.
        """
        return super(BaseScriptableCommand, self).getArgsDescription()

    ### \internal, don't call these methods.
    def _declareArg_Python(self, key, flags, defaultValue):
        """ \internal, impl. of CppCommands.BaseScriptableCommand_Python. """
        error = super(BaseScriptableCommand, self)._declareArg_Python(
            key, flags, defaultValue)
        if error:
            raise Exception(error)  
        return error;

    def _isArg_Python(self, key, flag):
        """ \internal, impl. of CppCommands.BaseScriptableCommand_Python. """
        error, hasFlag = super(BaseScriptableCommand, self)._isArg_Python(
            key, flag)
        if error:
            raise Exception(error)  
        return error, hasFlag;

    def _getArg_Python(self, key):
        """ \internal, impl. of CppCommands.BaseScriptableCommand_Python. """
        error, arg = super(BaseScriptableCommand, self)._getArg_Python(key)
        if error:
            raise Exception(error)  
        return error, arg;

    def _setArg_Python(self, key, arg):
        """ \internal, impl. of CppCommands.BaseScriptableCommand_Python. """
        error = super(BaseScriptableCommand, self)._setArg_Python(
            key, arg)
        if error:
            raise Exception(error)  
        return error;

    def _validateSetArgs_Python(self):
        """ \internal, impl. of CppCommands.BaseScriptableCommand_Python. """
        error = super(BaseScriptableCommand, self)._validateSetArgs_Python()
        if error:
            raise Exception(error)  
        return error;